src/main/java/com/snowflake/kafka/connector/internal/streaming/schemaevolution/TableSchemaResolver.java

```diff
@@ -30,40 +32,71 @@ protected TableSchemaResolver(ColumnTypeMapper columnTypeMapper) {
-        } else {
-          // Get from the schema
-          columnInfos = schemaMap.get(field.getKey());
-          if (columnInfos == null) {
-            // only when the type of the value is unrecognizable for JAVA
-            throw SnowflakeErrors.ERROR_5022.getException(
-                "column: " + field.getKey() + " schemaMap: " + schemaMap);
-          }
-        }
-        columnToType.put(colName, columnInfos);
-      }
+    Set<String> columnNamesSet = new HashSet<>(columnsToInclude);
+
+    if (hasSchema(record)) {
+      return getTableSchemaFromRecordSchema(recordNode, columnNamesSet, record);
+    } else {
+      return getTableSchemaFromJson(recordNode, columnNamesSet);
     }
-    return new TableSchema(columnToType);
+  }
+
+  private boolean hasSchema(SinkRecord record) {
+    return record.valueSchema() != null
+        && record.valueSchema().fields() != null
+        && !record.valueSchema().fields().isEmpty();
+  }
+
+  private TableSchema getTableSchemaFromRecordSchema(
+      JsonNode recordNode, Set<String> columnNamesSet, SinkRecord record) {
+    Map<String, ColumnInfos> schemaMap = getFullSchemaMapFromRecord(record);
+    Map<String, ColumnInfos> columnsInferredFromSchema =
+        Streams.stream(recordNode.fields())
+            .map(ColumnValuePair::of)
+            .filter(pair -> columnNamesSet.contains(pair.getQuotedColumnName()))
+            .peek(
+                field -> {
+                  if (!schemaMap.containsKey(field.getColumnName())) {
+                    // only when the type of the value is unrecognizable for JAVA
+                    throw SnowflakeErrors.ERROR_5022.getException(
+                        "column: " + field.getColumnName() + " schemaMap: " + schemaMap);
+                  }
+                })
+            .map(
+                field ->
+                    Maps.immutableEntry(
+                        Utils.quoteNameIfNeeded(field.getQuotedColumnName()),
+                        schemaMap.get(field.getColumnName())))
+            .collect(
+                Collectors.toMap(
+                    Map.Entry::getKey, Map.Entry::getValue, (oldValue, newValue) -> newValue));
+    return new TableSchema(columnsInferredFromSchema);
+  }
+
+  private TableSchema getTableSchemaFromJson(JsonNode recordNode, Set<String> columnNamesSet) {
+    Map<String, ColumnInfos> columnsInferredFromJson =
+        Streams.stream(recordNode.fields())
+            .map(ColumnValuePair::of)
+            .filter(pair -> columnNamesSet.contains(pair.getQuotedColumnName()))
+            .map(
+                pair ->
+                    Maps.immutableEntry(
+                        pair.getQuotedColumnName(),
+                        new ColumnInfos(inferDataTypeFromJsonObject(pair.getJsonNode()))))
+            .collect(
+                Collectors.toMap(
+                    Map.Entry::getKey, Map.Entry::getValue, (oldValue, newValue) -> newValue));
```

Comment:
open to suggestions how to make the refactoring nicer, java 8 does not help, kotlin would be perfect here ;(

Reply:
in theory we could split the schemaFromJson and schemaFromRecordSchema into two separate classes but i wouldn't invest here much more as this class is likely to be fully changed for the complex iceberg types